### ISSUE TITLE
 Initialization error "wrong-type-argument keymapp"

### ISSUE THREAD

**Issue by wanderley** [greghendershott/racket-mode#385]
I noticed the following error after update my `racket-mode` (which is AWESOME!)

```
Debugger entered--Lisp error: (wrong-type-argument keymapp ((keymap (91 . racket-smart-open-bracket) keymap (3 keymap (27 keymap (12 . paredit-recenter-on-sexp))) (C-M-right . paredit-backward-barf-sexp) (67108987 . paredit-backward-barf-sexp) (C-M-left . paredit-backward-slurp-sexp) (67108904 . paredit-backward-slurp-sexp) (C-left . paredit-forward-barf-sexp) (67108989 . paredit-forward-barf-sexp) (C-right . paredit-forward-slurp-sexp) (67108905 . paredit-forward-slurp-sexp) (M-down . paredit-splice-sexp-killing-forward) (M-up . paredit-splice-sexp-killing-backward) (11 . paredit-kill) (127 . paredit-backward-delete) (deletechar . paredit-forward-delete) (delete . paredit-forward-delete) (4 . paredit-forward-delete) (10 . paredit-newline) (59 . paredit-semicolon) (92 . paredit-backslash) (34 . paredit-doublequote) (93 . paredit-close-square) (91 . paredit-open-square) (27 keymap (113 . paredit-reindent-defun) (74 . paredit-join-sexps) (83 . paredit-split-sexp) (C-right . paredit-backward-barf-sexp) (C-left . paredit-backward-slurp-sexp) (63 . paredit-convolute-sexp) (114 . paredit-raise-sexp) (down . paredit-splice-sexp-killing-forward) (up . paredit-splice-sexp-killing-backward) (115 . paredit-splice-sexp) (40 . paredit-wrap-round) (14 . paredit-forward-up) (16 . paredit-backward-down) (4 . paredit-forward-down) (21 . paredit-backward-up) (2 . paredit-backward) (6 . paredit-forward) (127 . paredit-backward-kill-word) (100 . paredit-forward-kill-word) (59 . paredit-comment-dwim) (34 . paredit-meta-doublequote) (41 . paredit-close-round-and-newline)) (41 . paredit-close-round) (40 . paredit-open-round))))
  substitute-command-keys("Memory exhausted--use \\[save-some-buffers] then exit and restart Emacs")
  command-line()
  normal-top-level()
```

The error is gone if I add the following lines from the docs:
```
(use-package paredit
  :ensure t
  :config
  (dolist (m '(emacs-lisp-mode-hook
	       racket-mode-hook
	       racket-repl-mode-hook))
    (add-hook m #'paredit-mode))
  (bind-keys :map paredit-mode-map
	     ("{"   . paredit-open-curly)
	     ("}"   . paredit-close-curly))
  (unless terminal-frame
    (bind-keys :map paredit-mode-map
	       ("M-[" . paredit-wrap-square)
	       ("M-{" . paredit-wrap-curly))))
```

<details>

```
((alist-get 'racket-mode package-alist)
 #s(package-desc racket-mode
                 (20190621 1532)
                 "Major mode for Racket language."
                 ((emacs
                   (24 3))
                  (faceup
                   (0 0 2)))
                 nil nil "/Users/wander/.emacs.d/elpa/racket-mode-20190621.1532"
                 ((:url . "https://www.racket-mode.com/")
                  (:maintainer "Greg Hendershott")
                  (:authors
                   ("Greg Hendershott")))
                 nil))
((emacs-version "26.2")
 (emacs-uptime "8 seconds")
 (system-type darwin)
 (major-mode debugger-mode)
 (racket--el-source-dir "/Users/wander/.emacs.d/elpa/racket-mode-20190621.1532/")
 (racket--rkt-source-dir "/Users/wander/.emacs.d/elpa/racket-mode-20190621.1532/racket/")
 (racket-program "racket")
 (racket-command-port 55555)
 (racket-command-startup nil)
 (racket-command-timeout 10)
 (racket-memory-limit 2048)
 (racket-error-context medium)
 (racket-retry-as-skeleton t)
 (racket-error-context medium)
 (racket-history-filter-regexp "\\`\\s *\\S ?\\S ?\\s *\\'")
 (racket-images-inline t)
 (racket-images-keep-last 100)
 (racket-use-repl-submit-predicate nil)
 (racket-images-system-viewer "display")
 (racket-pretty-print t)
 (racket-indent-curly-as-sequence t)
 (racket-indent-sequence-depth 0)
 (racket-pretty-lambda nil)
 (racket-smart-open-bracket-enable nil)
 (racket-module-forms "\\s(\\(?:module[*+]?\\|library\\)")
 (racket-logger-config
  ((cm-accomplice . warning)
   (GC . info)
   (module-prefetch . warning)
   (optimizer . info)
   (racket/contract . error)
   (sequence-specialization . info)
   (* . fatal))))
(enabled-minor-modes
 (async-bytecomp-package-mode)
 (auto-composition-mode)
 (auto-compression-mode)
 (auto-encryption-mode)
 (auto-fill-mode)
 (auto-save-mode)
 (blink-cursor-mode)
 (buffer-read-only)
 (column-number-mode)
 (company-mode)
 (diff-auto-refine-mode)
 (electric-pair-mode)
 (file-name-shadow-mode)
 (font-lock-mode)
 (global-auto-revert-mode)
 (global-company-mode)
 (global-eldoc-mode)
 (global-font-lock-mode)
 (global-git-commit-mode)
 (global-magit-file-mode)
 (global-origami-mode)
 (ivy-mode)
 (line-number-mode)
 (magit-auto-revert-mode)
 (mouse-wheel-mode)
 (origami-mode)
 (override-global-mode)
 (projectile-mode)
 (savehist-mode)
 (shackle-mode)
 (shell-dirtrack-mode)
 (show-paren-mode)
 (transient-mark-mode)
 (which-key-mode)
 (winner-mode))
(disabled-minor-modes
 (abbrev-mode)
 (auto-fill-function)
 (auto-revert-mode)
 (auto-revert-tail-mode)
 (auto-save-visited-mode)
 (cl-old-struct-compat-mode)
 (company-search-mode)
 (compilation-in-progress)
 (compilation-minor-mode)
 (compilation-shell-minor-mode)
 (completion-in-region-mode)
 (counsel-mode)
 (counsel-projectile-mode)
 (defining-kbd-macro)
 (delete-selection-mode)
 (diff-minor-mode)
 (dired-hide-details-mode)
 (display-time-mode)
 (edebug-mode)
 (eldoc-mode)
 (electric-indent-mode)
 (electric-layout-mode)
 (electric-quote-mode)
 (elm-format-on-save-mode)
 (elm-indent-mode)
 (flymake-mode)
 (git-commit-mode)
 (git-gutter-mode)
 (global-git-gutter-mode)
 (global-prettify-symbols-mode)
 (global-visual-line-mode)
 (hg-sl-mode)
 (hl-sentence-mode)
 (horizontal-scroll-bar-mode)
 (hs-minor-mode)
 (ibuffer-auto-mode)
 (ido-everywhere)
 (isearch-mode)
 (jit-lock-debug-mode)
 (lsp--managed-mode)
 (lsp-lens-mode)
 (lsp-mode)
 (magit-blame-mode)
 (magit-blame-read-only-mode)
 (magit-blob-mode)
 (magit-file-mode)
 (magit-wip-after-apply-mode)
 (magit-wip-after-save-local-mode)
 (magit-wip-after-save-mode)
 (magit-wip-before-change-mode)
 (magit-wip-initial-backup-mode)
 (magit-wip-mode)
 (mail-abbrevs-mode)
 (markdown-live-preview-mode)
 (mc-hide-unmatched-lines-mode)
 (menu-bar-mode)
 (mml-mode)
 (monky-blame-mode)
 (monky-branches-mode)
 (monky-commit-mode)
 (monky-log-mode)
 (monky-queue-mode)
 (monky-status-mode)
 (multiple-cursors-mode)
 (next-error-follow-minor-mode)
 (org-cdlatex-mode)
 (org-src-mode)
 (outline-minor-mode)
 (outorg-edit-minor-mode)
 (outshine-mode)
 (overwrite-mode)
 (paragraph-indent-minor-mode)
 (paredit-mode)
 (prettify-symbols-mode)
 (pyvenv-mode)
 (pyvenv-tracking-mode)
 (racket-check-syntax-mode)
 (rectangle-mark-mode)
 (rectangular-region-mode)
 (save-place-mode)
 (server-mode)
 (sh-electric-here-document-mode)
 (shell-command-with-editor-mode)
 (size-indication-mode)
 (smerge-mode)
 (temp-buffer-resize-mode)
 (tool-bar-mode)
 (tooltip-mode)
 (transient-resume-mode)
 (unify-8859-on-decoding-mode)
 (unify-8859-on-encoding-mode)
 (url-handler-mode)
 (use-hard-newlines)
 (view-mode)
 (visible-mode)
 (visual-line-mode)
 (which-function-mode)
 (window-divider-mode)
 (with-editor-mode)
 (xref-etags-mode)
 (yas-global-mode)
 (yas-minor-mode))
```

</details>

**Comment by wanderley** [comment]
I think that this issue is related with `use-package` because the following works fine
```
(use-package racket-mode :ensure t)
(use-package paredit :ensure t)
```
but
```
(use-package racket-mode :ensure t)
(use-package paredit
  :ensure t
  :hook (emacs-lisp-mode . paredit-mode))
```
raises the same error.

**Comment by greghendershott** [comment]
Sorry about that; thank you for reporting this. It's got to be related to commit a19dba1 from a couple weeks ago which was supposed to fix #378. I'll take a look and try to fix this soon.

**Comment by greghendershott** [comment]
@wanderley By the way, I'd never heard of `:hook` until you mentioned it. I double-checked the `use-package` doc string and didn't see it mentioned or explained.

Taking a quick/shallow look in `use-package.el`, I do see `:hook`. Indeed it is a thing.

But. Do you know:
- Is it maybe "experimental"?
- How/when it is suppose to be used instead of using `add-hook`?

**Comment by greghendershott** [comment]
I found time to look at this more carefully. Update:

- I learned that although `:hook` isn't in the doc string it _is_ in the `use-package` README and Info documentation.

- Although I still don't understand exactly why this error is happening, it seems related to load order. And the more I think about this, the more I think Racket Mode is trying to be too helpful and automagical -- and not very "Emacsy".

  Currently we _always_ try to rebind <kbd>[</kbd>, then look at the value of the `racket-smart-open-bracket-enable` customization variable.

  Instead: I think we should ask people to do the usual Emacs thing: If you want a key to do something different, then bind the key some different command.  So: If you want to use this feature, then bind the <kbd>[</kbd> key to the `racket-smart-open-bracket` command in the `racket-mode-map` and/or `racket-repl-mode-map` keymaps. Else if you don't, then don't.

  The `racket-smart-open-bracket-enable` customization variable will become obsolete. Racket Mode will no longer attempt to detect paredit and modify its keymap automagically -- and as a result won't encounter whatever load-order problem is happening here.

  I have some code to do this, and, it seems like "it works". I'll give it a longer think and some more testing before I push a commit.

**Comment by wanderley** [comment]
>  Although I still don't understand exactly why this error is happening, it seems related to load order. And the more I think about this, the more I think Racket Mode is trying to be too helpful and automagical -- and not very "Emacsy".

Were you able to repro it?

> Instead: I think we should ask people to do the usual Emacs thing: If you want a key to do something different, then bind the key some different command. So: If you want to use this feature, then bind the [ key to the racket-smart-open-bracket command in the racket-mode-map and/or racket-repl-mode-map keymaps. Else if you don't, then don't.

This behavior is more emacsy, and one might prefer to add a configuration to enable a feature than adding a configuration to prevent a load-order issue.

> I have some code to do this, and, it seems like "it works". I'll give it a longer think and some more testing before I push a commit.

Sure. I am using the suggestion from the docs for a while.

**Comment by greghendershott** [comment]
@wanderley Yes. I was able to repro it with the status quo. And to confirm the new approach does not cause the same error -- at least for me.

I wish I could tell you exactly how and why the `use-package` `:hook` code causes this problem but a manual `add-hook` doesn't. If I could explain that, exactly, I'd feel more confident and probably already have merged commit 669275b to master.

Also, I needed to revise some doc strings, and I've found it pays to re-read my documentation prose the next day to make it suck less. :smile:

Anyway I should probably go ahead and merge, today.
